python/tests/test_ld_stats.py

```diff
@@ -0,0 +1,475 @@
+# MIT License
+#
+# Copyright (c) 2018-2019 Tskit Developers
+# Copyright (C) 2016 University of Oxford
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Test cases for ld matrix calculations.
+"""
+import unittest
+import io
+
+import numpy as np
+import msprime
+
+import tskit
+import _tskit
+#import tests.tsutil as tsutil
+#import tests.test_wright_fisher as wf
+
+
+##############################
+# LD matrix calculator
+##############################
+
+
+## notes 6/12:
+## Question: Do we want to return the LD matrix over variants or over sites?
+##  I think sites, and that's what the naive_ld_matrix returns
+##
+## Because the returned LD matrix might not include every variant, and variants may be
+##  distributed over multiple windows, we probably need to also return information of
+##  which index in each LD matrix corresponds to which mutation in the tree sequence.
+##  Thus, probably want to also return list of lists of site id indexes.
+##
+## Should windows be a list of non-overlapping intervals? Previously, I set up so that we
+##  just define window edges, so all windows necessarily have zero separation. Now we
+##  instead are able to compute LD matrices or LD scores from windows that are
+##  not adjacent. So could look like `windows = [(100, 200), (500, 1000), (1000, 2000)]`
```

Comment:
Oh, this is nice way of allowing either - how about if `windows` is just a single flat list, it is then interpreted as being the breakpoints between contiguous windows (that must span the whole genome) and converted to this form?

Should `windows` refer to both the rows *and* columns of the LD matrix? Or just one?

Reply:
Oh, I see you're only doing the diagonal blocks as specified by the windows. Makes sense. But, should offdiagonals be allows? Would people want to compute, say, LD between a given SNP and all others in the genome?

Thinking a bit more about this - I think the LD matrix shold have `windows="sites"`, for consistency? And then be given another argument to specify *which* pairs of sites to output, if not all of them. I would have thought that we'd want something like `max_distance`, to get the diagonal band of the matrix, rather than filling out only blocks along the diagonal as this does (IIUC).  If you pass in different windows (ie, numbers, not "sites"), you'd get an nwindows x nwindows matrix whose (i,j)th entry is the average LD between SNPs in windows i and j (or, the corresponding thing integrated over the trees).

Reply:
There are a few ways to set it up - and I agree that it would be nice to be more flexible (say, compute LD between this set and that set of sites or windows, like you say). I don't know if we want to require that windows cover the entire sequence length. Unlike the single site stats, two-locus stats are more expensive to compute, and will get more expensive to store a big N sites by M sites matrix.

I do like the idea of setting some focal window length to only get LD between pairs of variants within some distance, giving you a diagonal band.

There are definitely some more choices to be made about generalizing and argument options. These points you bring up are important to think about..